### ISSUE TITLE
Bug: [prefer-nullish-coalescing] internal error when linting on binary expression (||) with nullable type

### ISSUE THREAD

**Issue by JLHwung** [typescript-eslint/typescript-eslint#7222]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play#ts=5.1.6&fileType=.tsx&code=DYUwLgBAhgXBDOYBOBLAdgcwgHwmgrsMBALx6HADcAUKJAEZyKqY7lGntXXVQ671qQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WlqYSNkAC1pkA9gEMkyMswDm6KL2jjokcGAC%2BILUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```typescript
let a: string | null = null;
let b: string | null = null;

a || b
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/prefer-nullish-coalescing": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

No internal errors

### Actual Result

eslint throws when linting on the line `a || b`
```
TypeError: Cannot read properties of undefined (reading 'some')
Occurred while linting /input.tsx:4
Rule: "@typescript-eslint/prefer-nullish-coalescing"
    at LogicalExpression[operator = "||"] (https://typescript-eslint.io/sandbox/index.js:140:49356)
    at https://typescript-eslint.io/sandbox/index.js:105:2536
    at https://typescript-eslint.io/sandbox/index.js:99:32978
    at Array.forEach (<anonymous>)
    at Object.emit (https://typescript-eslint.io/sandbox/index.js:99:32967)
    at Sx.applySelector (https://typescript-eslint.io/sandbox/index.js:40:22861)
    at Sx.applySelectors (https://typescript-eslint.io/sandbox/index.js:40:23157)
    at Sx.enterNode (https://typescript-eslint.io/sandbox/index.js:40:23248)
    at ex.enterNode (https://typescript-eslint.io/sandbox/index.js:34:23858)
    at https://typescript-eslint.io/sandbox/index.js:105:2972 4:1 - 4:7
```

The error is probably thrown from

https://github.com/typescript-eslint/typescript-eslint/blob/058c449935cf209d433b190fa5b65a9614935d7b/packages/eslint-plugin/src/rules/prefer-nullish-coalescing.ts#L313

### Additional Info

(I didn't figure out how to run the lint again with `--debug` flag on the playground)

The reproduction is extracted from the Babel source: https://github.com/babel/babel/blob/main/packages/babel-helper-create-class-features-plugin/src/features.ts, first reported [here](https://github.com/babel/babel/pull/15716#issuecomment-1633362270).
